chips/pc/src/pit.rs

```diff
@@ -0,0 +1,195 @@
+//! unavoidable due to hardware limitation. For instance, the PIT's internal frequency is a rational
+//! number but not an integer, so all calculations are imprecise due to integer rounding.
+//!
+//! This implementation is based on guidance from the following sources:
+//!
+//! * https://wiki.osdev.org/Programmable_Interval_Timer
+//! * https://en.wikipedia.org/wiki/Intel_8253
+
+use core::cell::Cell;
+
+use kernel::hil::time::{Alarm, AlarmClient, Frequency, Ticks, Ticks32, Time};
+use kernel::ErrorCode;
+use tock_cells::numeric_cell_ext::NumericCellExt;
+use tock_cells::optional_cell::OptionalCell;
+use x86::io;
+
+/// Frequency of the PIT's internal oscillator
+///
+/// According to Wikipedia, this should be "one third of the NTSC color subcarrier frequency."
+///
+/// Why such a specific value? This frequency is heavily used in analog television circuitry. At the
+/// time the 8253 was first introduced, it was very easy and cheap to obtain a crystal oscillator at
+/// exactly this frequency.
+const OSCILLATOR_FREQUENCY: u32 = 3579545 / 3;
+
+/// Frequency of the PIT timer
+///
+/// Parameter `R` is the PIT's reload value for channel 0.
+///
+/// Internal oscillator frequency is a rational, non-integer number. This means there is some loss
+/// of precision because Tock uses integers to represent frequency.
+pub struct PitFreq<const R: u16>;
+
+impl<const R: u16> Frequency for PitFreq<R> {
+    fn frequency() -> u32 {
+        OSCILLATOR_FREQUENCY / (R as u32)
+    }
+}
+
+/// Computes a PIT reload value for the given frequency.
+///
+/// The actual interrupt frequency will always be slightly different from the requested `freq` due
+/// to hardware limitations. This function tries to get as close as possible.
+pub const fn reload_value(freq: u32) -> u16 {
+    // Lowest possible frequency is about 18 Hz
+    if freq <= 18 {
+        // PIT interprets zero as "maximum possible value"
+        return 0x0000;
+    }
+
+    // Highest possible frequency is OSCILLATOR_FREQUENCY
+    if freq >= OSCILLATOR_FREQUENCY {
+        return 0x0001;
+    }
+
+    let mut r = OSCILLATOR_FREQUENCY / freq;
+
+    // If remainder is more than half, then round up to get as close as possible
+    if (OSCILLATOR_FREQUENCY % freq) >= (OSCILLATOR_FREQUENCY / 2) {
+        r += 1;
+    }
+
+    r as u16
+}
+
+/// Reload value corresponding to 1 KHz
+pub const RELOAD_1KHZ: u16 = reload_value(1000);
+
+/// I/O port address for channel 0
+const PIT_CD0: u16 = 0x0040;
+
+/// I/O port address for mode/command register
+const PIT_MCR: u16 = 0x0043;
+
+/// Specifies channel zero when writing to the mode/command register
+const PIT_MCR_C0: u8 = 0b00000000;
+
+/// Specifies lobyte/hibyte access mode when writing to the mode/command register
+const PIT_MCR_LOHI: u8 = 0b00110000;
+
+/// Specifies mode 2 (rate generator) when writing to the mode/command register
+const PIT_MCR_M2: u8 = 0b00000100;
```

Comment:
Maybe we can define the registers and =use `LocalRegisterCopy`?

Reply:
Latest push refactors this driver to use `LocalRegisterCopy`